### ISSUE TITLE
samples with amqpalloc_set_memory_tracing_enabled(false) causes crash

### ISSUE THREAD

**Issue by SparhawkSoftware** [Azure/azure-uamqp-c#13]
Is it supposed to work without memory tracing?

**Comment by dcristoloveanu** [comment]
Let me have a quick look at this and I will come back with what's going on.

**Comment by SparhawkSoftware** [comment]
any information.
I created a new issue because now trace_free is crashing as well.

**Comment by dcristoloveanu** [comment]
I finally got to this.

I just tried to run message_sender_sample with amqpalloc_set_memory_tracing_enabled(false); on Windows and it seems to run.
Can you give me the details of what you are running and on what platform so I can repro?

Thanks,
/Dan

**Comment by SparhawkSoftware** [comment]
The trace above shows up in my app and in the sample programs.
I am using a build with OpenSSL enabled.

**Comment by dcristoloveanu** [comment]
I switched to using OpenSSL on Windows after I pulled in the changes to OpenSSL in C shared utility. I'm hitting a different issue where it seems the handshake for tlsio_openssl does not complete, once I'm done with that I'll see where this AV comes from.

I'm using message_sender_sample as this is the simplest and based on the above I suspect this should be enough to repro it.

**Comment by dcristoloveanu** [comment]
@SparhawkSoftware 

Hi Randy,

I figured my issue was due to the way I was doing cert validation in my sample code.
I got to try your receiver sample now with OpenSSL on Windows without memory tracing.

Here is what I did:
- Took your sample code (I pushed it to this branch for reference: https://github.com/Azure/azure-uamqp-c/tree/dcristo_repro_issue_13)
- cmake .. -Duse_openssl:bool=ON -Dmemory_trace:bool=OFF
- Open the uamqp.sln
- Went to main.c in the message_receiver_sample and placed some EH credentials in there (the EH name, the key name and the actual key)
- Ran the sample and was able to receive data without any segfault/AV.

I am probably doing something wrong and that's why I suspect I can't repro, can you have a look at these steps and tell me if I am missing anything?

Cheers,
/Dan
P.S. I also tried with amqpalloc_set_memory_tracing_enabled(false); and also no repro.

**Comment by dcristoloveanu** [comment]
I forgot to mention I am using openssl 1.0.1t.

**Comment by SparhawkSoftware** [comment]
I can't replicate with your branch because you did not pick up the updates I added for OpenSSL.

These updates are now integrated into the development branch here:
https://github.com/Azure/azure-c-shared-utility

I tried your sample with the develop branch and could replicate.
I used my message_sender_sample from here:
https://github.com/SparhawkSoftware/azure-uamqp-c

I could replicate the problem.
So it appears that it is an issue with the changes I made.
Note that I am not using EventHub. 
I am using simple AMQP queues with Azure and ActiveMQ

I can give you access to my brokers if you still can't replicate.

**Comment by SparhawkSoftware** [comment]
I also have the same problem with SSPI enabled.

**Comment by SparhawkSoftware** [comment]
I removed the code I added to set the OpenSSL validation callback and I can still replicate.

(e.g. I added a dummy function inside of tlsio_openssl.c to validate and commented out the calls to xio_setoption)

At this point my theory is it is related to the fact that I am not connecting to an EH and I changed the value that was passed into messaging_create_target (I can't change it back since I can't connect using the original logic).

**Comment by dcristoloveanu** [comment]
@SparhawkSoftware If you could give me access and push your branch exactly as is I can take a look and see what's going on so we nail this down.

**Comment by SparhawkSoftware** [comment]
You should have access.

**Comment by dcristoloveanu** [comment]
Ok, got it. I assume this would be the master branch on your fork.
I'll try that with a SB queue.
In the meanwhile if you can e-mail me (dcristo@microsoft.com) an endpoint for a broker that you use I can try against that directly also.

Thanks for the patience on this.
/Dan

**Comment by dcristoloveanu** [comment]
@SparhawkSoftware 
I reproed the issue with your branch (happens with the broker you gave me and no memory tracing).
It was a wrong double free which was not manifesting with the memory tracing.
The offending line is link.c, ln. 346 in your branch.

This has been fixed on Aug 9 per git log.

I can also see that your fork and branch are not using the latest changes I pulled in from you for the openssl.

I think if you would merge with the latest develop branch for uAMQP the crash will go away and you'll be able to use the use_openssl option you contributed.

Let me know if there's still any issue.

Cheers,
/Dan
